src/openrct2/scripting/ScriptEngine.cpp

```diff
@@ -0,0 +1,956 @@
+            }
+            _ss << " ]";
+            PopIndent(2);
+            if (canStartWithNewLine)
+            {
+                PopIndent();
+            }
+        }
+        duk_pop(_context);
+    }
+
+    void StringifyObject(const DukValue& val, bool canStartWithNewLine)
+    {
+        auto numEnumerables = GetNumEnumerablesOnObject(val);
+        if (numEnumerables == 0)
+        {
+            _ss << "{}";
+        }
+        else if (numEnumerables == 1)
+        {
+            _ss << "{ ";
+
+            val.push();
+            duk_enum(_context, -1, 0);
+            auto index = 0;
+            while (duk_next(_context, -1, 1))
+            {
+                if (index != 0)
+                {
+                    _ss << ", ";
+                }
+                auto value = DukValue::take_from_stack(_context, -1);
+                auto key = DukValue::take_from_stack(_context, -1);
+                if (key.type() == DukValue::Type::STRING)
+                {
+                    _ss << key.as_string() << ": ";
+                }
+                else
+                {
+                    // For some reason the key was not a string
+                    _ss << "?: ";
+                }
+                Stringify(value, true);
+                index++;
+            }
+            duk_pop_2(_context);
+
+            _ss << " }";
+        }
+        else
+        {
+            if (canStartWithNewLine)
+            {
+                PushIndent();
+                LineFeed();
+            }
+
+            _ss << "{ ";
+            PushIndent(2);
+
+            val.push();
+            duk_enum(_context, -1, 0);
+            auto index = 0;
+            while (duk_next(_context, -1, 1))
+            {
+                if (index != 0)
+                {
+                    _ss << ",";
+                    LineFeed();
+                }
+                auto value = DukValue::take_from_stack(_context, -1);
+                auto key = DukValue::take_from_stack(_context, -1);
+                if (key.type() == DukValue::Type::STRING)
+                {
+                    _ss << key.as_string() << ": ";
+                }
+                else
+                {
+                    // For some reason the key was not a string
+                    _ss << "?: ";
+                }
+                Stringify(value, true);
+                index++;
+            }
+            duk_pop_2(_context);
+
+            PopIndent(2);
+            _ss << " }";
+
+            if (canStartWithNewLine)
+            {
+                PopIndent();
+            }
+        }
+    }
+
+    void StringifyFunction(const DukValue& val)
+    {
+        val.push();
+        if (duk_is_c_function(_context, -1))
+        {
+            _ss << "[Native Function]";
+        }
+        else if (duk_is_ecmascript_function(_context, -1))
+        {
+            _ss << "[ECMAScript Function]";
+        }
+        else
+        {
+            _ss << "[Function]";
```

Comment:
if not native and not ecma, then what could this be?

Reply:
duktape has concept of lightfunc which I don't fully understand or know how to produce. This is defensive programming to make sure something is output for any value.